src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +27,141 @@ class DictionarySegment : public BaseSegment {
+  void append(const AllTypeVariant& val) override {
+    throw std::runtime_error("Append failed: DictionarySegment is immutable.");
+  }
 
   // returns an underlying dictionary
-  std::shared_ptr<const std::vector<T>> dictionary() const;
+  std::shared_ptr<const std::vector<T>> dictionary() const { return _dictionary; }
 
   // returns an underlying data structure
-  std::shared_ptr<const BaseAttributeVector> attribute_vector() const;
+  std::shared_ptr<BaseAttributeVector> attribute_vector() const { return _attribute_vector; }
 
   // return the value represented by a given ValueID
-  const T& value_by_value_id(ValueID value_id) const;
+  const T& value_by_value_id(ValueID value_id) const { return _dictionary->at(value_id); }
 
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    auto forward_iterator = std::lower_bound(_dictionary->begin(), _dictionary->end(), value);
+    if (forward_iterator == _dictionary->end()) {
+      return INVALID_VALUE_ID;
+    }
+    return (ValueID)(forward_iterator - _dictionary->begin());
+  }
 
   // same as lower_bound(T), but accepts an AllTypeVariant
-  ValueID lower_bound(const AllTypeVariant& value) const;
+  ValueID lower_bound(const AllTypeVariant& value) const { return lower_bound(value); }
 
   // returns the first value ID that refers to a value > the search value
   // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
-  ValueID upper_bound(T value) const;
+  ValueID upper_bound(T value) const {
+    auto forward_iterator = std::upper_bound(_dictionary->begin(), _dictionary->end(), value);
+    if (forward_iterator == _dictionary->end()) {
+      return INVALID_VALUE_ID;
+    }
+    return (ValueID)(forward_iterator - _dictionary->begin());
+  }
 
   // same as upper_bound(T), but accepts an AllTypeVariant
-  ValueID upper_bound(const AllTypeVariant& value) const;
+  ValueID upper_bound(const AllTypeVariant& value) const { return upper_bound(value); }
 
   // return the number of unique_values (dictionary entries)
-  size_t unique_values_count() const;
+  size_t unique_values_count() const { return _dictionary->size(); }
 
   // return the number of entries
-  ChunkOffset size() const override;
+  ChunkOffset size() const override { return _attribute_vector->size(); }
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+    return _dictionary->size() * sizeof(T) + _attribute_vector->size() * _attribute_vector->width();
+  }
 
  protected:
   std::shared_ptr<std::vector<T>> _dictionary;
   std::shared_ptr<BaseAttributeVector> _attribute_vector;
-};
 
+  std::shared_ptr<std::vector<T>> _initialize_dictionary(const std::shared_ptr<ValueSegment<T>> value_segment) {
+    std::set<T> distinct_values;
```

Comment:
As discussed in the lecture, it could be worth it to use `unordered_set` instead of `set` performancewise.